### PR TITLE
update delete handler to use model.Name

### PR DESCRIPTION
**Pull request by mxyng**
also update list handler to use the `Manifests()` iter

follow up to #3715

### CODE REVIEW

**Review by BruceMacD — APPROVED** [review]

File: server/layer.go [hunk 1]
```diff
@@ -88,3 +90,27 @@ func (l *Layer) Open() (io.ReadCloser, error) {
 
 	return os.Open(blob)
 }
+
+func (l *Layer) Remove() error {
+	var count int
+	Manifests()(func(name model.Name, manifest *Manifest) bool {
+		for _, layer := range append(manifest.Layers, manifest.Config) {
+			if layer.Digest == l.Digest {
+				count++
```
Comment (mxyng) [inline server/layer.go:99]: can probably short circuit this and return false

File: server/layer.go [hunk 2]
```diff
@@ -88,3 +90,27 @@ func (l *Layer) Open() (io.ReadCloser, error) {
 
 	return os.Open(blob)
 }
+
+func (l *Layer) Remove() error {
+	var count int
+	Manifests()(func(name model.Name, manifest *Manifest) bool {
+		for _, layer := range append(manifest.Layers, manifest.Config) {
```
Comment (mxyng) [inline server/layer.go:97]: maybe build a map with the digests so it doesn't need to loop every time

File: server/routes.go [hunk 3]
```diff
@@ -573,48 +573,31 @@ func (s *Server) CreateModelHandler(c *gin.Context) {
...
-		return
-	}
-
-	if err := DeleteModel(model); err != nil {
-		if os.IsNotExist(err) {
-			c.JSON(http.StatusNotFound, gin.H{"error": fmt.Sprintf("model '%s' not found", model)})
-		} else {
-			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
-		}
+	n := model.ParseName(cmp.Or(r.Model, r.Name))
```
Comment (jmorganca) [inline server/routes.go:585]: Out of scope for this PR but it seems this should probably return an error:

```
m, err := model.Parse((cmp.Or(r.Model, r.Name))
if err != nil {
  ...
}
```

File: server/manifest_test.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+			filepath.Join("host", "namespace", "model", "tag"),
+			filepath.Join("host", "namespace", "model", ".hidden"),
+		},
+		"subdir": {
+			filepath.Join("host", "namespace", "model", "tag", "one"),
+			filepath.Join("host", "namespace", "model", "tag", "another", "one"),
+		},
+	}
+
+	for n, wants := range cases {
```
Comment (jmorganca) [inline server/manifest_test.go:65]: nice tests!

File: server/manifest.go [hunk 5]
```diff
@@ -25,30 +29,59 @@ func (m *Manifest) Size() (size int64) {
...
+	manifests, err := GetManifestPath()
+	if err != nil {
+		return err
+	}
+
+	return PruneDirectory(manifests)
+}
+
+func ParseNamedManifest(n model.Name) (*Manifest, error) {
+	if !n.IsFullyQualified() {
```
Comment (jmorganca) [inline server/manifest.go:52]: What is Qualified in this case? Does this mean it has all its components (vs just `llama3`) - wouldn't that always be the case if we have defaults?

Comment (jmorganca) [reply server/manifest.go:52]: In other words could we just catch this in `IsValid`?

File: server/manifest.go [hunk 6]
```diff
@@ -77,3 +110,48 @@ func WriteManifest(name string, config *Layer, layers []*Layer) error {
...
+		}
+
+		if !fi.IsDir() {
+			rel, err := filepath.Rel(manifests, match)
+			if err != nil {
+				slog.Warn("bad filepath", "path", match, "error", err)
+				continue
+			}
+
+			n := model.ParseNameFromFilepath(rel)
```
Comment (jmorganca) [inline server/manifest.go:140]: Will this risk any uppercase/lowercase comparison issues? (and should we fix that first?)

File: server/routes.go [hunk 7]
```diff
@@ -573,48 +573,31 @@ func (s *Server) CreateModelHandler(c *gin.Context) {
...
+	if !n.IsValid() {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("name %q is invalid", cmp.Or(r.Model, r.Name))})
 		return
 	}
 
-	manifestsPath, err := GetManifestPath()
+	m, err := ParseNamedManifest(n)
```
[the remark this answers is not in the document]
Comment (mxyng) [reply server/routes.go:591]: It's using model.Name to parse a Manifest

File: server/manifest.go (hunk 5 above)
Comment (mxyng) [reply server/manifest.go:52]: not necessarily. there's nothing stopping anyone from passing in an empty name

```go
ParseNamedManifest(model.Name{})
```

File: server/manifest.go (hunk 6 above)
Comment (mxyng) [reply server/manifest.go:140]: there's no transformation left in model.Name so the behaviour should be the same as now

**Pull request merged**